### PR TITLE
Update to Oculus SDK 0.8.0.0-beta

### PR DESCRIPTION
**Pull request by Squareys**
Hi @mosra !

This is an early PR to let you know I'm working on this. I actually found some old changes I forgot about in my local repository `:D`

TODOs:
 - [x] `ovr_Detect` function is back
 - [x] ovr_GetSessionStatus
 - [x] check error enum for hidden changes (yes, there were some)
 - [x] update cmake modules in examples
 - [x] investigate latencyMarker flag in ovrTrackingState
 - [x] ovr_ConfigureTracking is not required, but is there a use case where one would not want full tracking capabilities (just in case there is this use case, the method will be left, doc/example will be adapted)
 - [x] ovrLayerHudMode
 - Support for Oculus Touch will be prosponed until I can make use of it.

*(those TODOs were written late at night, my appologies for anything weird that may have resulted from that.)*

Greetings,
Squareys

### CODE REVIEW

**Comment by mosra** [comment]
Oh. Wow. Slow down, I can't keep up! :)

Now I have genuine fear that we get banned from the CIs soon :D

**Comment by Squareys** [comment]
I will try to batch more commits together before pushing, to grant the CIs a break every now and then.

**Comment by mosra** [comment]
I'm not complaining, it's just that the AppVeyor builds seemed to take three times longer than before (12 minutes vs. 4 minutes), like if we were throttled because of excessive activity :)

File: src/Magnum/OvrIntegration/Compositor.h [hunk 1]
```diff
@@ -27,7 +27,11 @@
 */
 
 /** @file
- * @brief Class @ref Magnum::OvrIntegration::Compositor, @ref Magnum::OvrIntegration::Layer, @ref Magnum::OvrIntegration::LayerDirect, @ref Magnum::OvrIntegration::LayerEyeFov, @ref Magnum::OvrIntegration::LayerEyeFovDepth, @ref Magnum::OvrIntegration::LayerQuad, @ref Magnum::OvrIntegration::TimewarpProjectionDescription, enum @ref Magnum::OvrIntegration::LayerType
+ * @brief Class @ref Magnum::OvrIntegration::Compositor, @ref Magnum::OvrIntegration::Layer,
+ *    @ref Magnum::OvrIntegration::HeadLockableLayer, @ref Magnum::OvrIntegration::LayerDirect,
+ *    @ref Magnum::OvrIntegration::LayerEyeFov, @ref Magnum::OvrIntegration::LayerEyeFovDepth,
+ *    @ref Magnum::OvrIntegration::LayerQuad, @ref Magnum::OvrIntegration::TimewarpProjectionDescription,
+ *    enum @ref Magnum::OvrIntegration::LayerType
```
Comment (mosra) [inline src/Magnum/OvrIntegration/Compositor.h]: Sorry, but this will break Doxygen, we *have to* put `@brief`s on a single line (annoying, I know).

File: src/Magnum/OvrIntegration/Context.h [hunk 2]
```diff
@@ -53,6 +53,17 @@ struct Error {
...
 /**
+ * @brief Result of @ref Context::detect(int).
+ */
+enum class OvrDetectResultFlag: Byte {
+    ServiceRunning = 1,
+    HmdConnected = 2
+};
+
+typedef Corrade::Containers::EnumSet<OvrDetectResultFlag> OvrDetectResult;
+CORRADE_ENUMSET_OPERATORS(OvrDetectResult)
```
Comment (mosra) [inline src/Magnum/OvrIntegration/Context.h]: Maybe naming this `OvrDetectResultFlags`? ... also, detect or detected?

File: src/Magnum/OvrIntegration/Hmd.cpp [hunk 3]
```diff
@@ -160,16 +160,39 @@ bool Hmd::isDebugHmd() const {
...
 
 void Hmd::setDebugHudStereoMode(const DebugHudStereoMode mode) const {
-    ovr_SetInt(_hmd, OVR_DEBUG_HUD_STEREO_MODE, Int(mode));
+    ovr_SetInt(_session, OVR_DEBUG_HUD_STEREO_MODE, Int(mode));
+}
+
+Hmd::EyeCup Hmd::eyeCup() const {
+    char value[16];
+    ovr_GetString(_session, OVR_KEY_EYE_CUP, "A");
+    if(strcmp(value, "A")) {
```
Comment (mosra) [inline src/Magnum/OvrIntegration/Hmd.cpp]: `std::strcmp`, please. I guess there is a reason why they are using string and not enum? For example, wouldn't this be vendor-dependent or something?

File: src/Magnum/OvrIntegration/Hmd.h [hunk 4]
```diff
@@ -191,6 +192,19 @@ Framebuffer::blit(mirrorFramebuffer,
...
+         * @see @ref Hmd::eyeCup()
+         */
+        enum class EyeCup: UnsignedShort {
+            A,
+            B,
+            C,
+            Unknown = 0xFF
+        };
+
+        /** Destructor. */
```
Comment (mosra) [inline src/Magnum/OvrIntegration/Hmd.h]:
```cpp
/**
 * @brief Eye cup
 * @see @ref eyeCup()
 */
```

The destructor doc is not needed as it doesn't bring any new information.

File: src/Magnum/OvrIntegration/Compositor.h (hunk 1 above)
Comment (Squareys) [inline src/Magnum/OvrIntegration/Compositor.h]: Alright, we probably had discussed that back then when the code was first written. I will change it back.

 - [ ]

File: src/Magnum/OvrIntegration/Context.h (hunk 2 above)
Comment (Squareys) [inline src/Magnum/OvrIntegration/Context.h]: Well it's the result of the `ovr_Detect` method, which makes this the result of Detect. See `static OvrDetectResult Context::detect(int)`

File: src/Magnum/OvrIntegration/Hmd.cpp (hunk 3 above)
Comment (Squareys) [inline src/Magnum/OvrIntegration/Hmd.cpp]: I don't think the Oculus SDK will be used by anybody else than Oculus, I can return string, of course, but it's always A, B or C. Who knows what will be in the future, though.

- [ ]

File: src/Magnum/OvrIntegration/Hmd.h (hunk 4 above)
Comment (Squareys) [inline src/Magnum/OvrIntegration/Hmd.h]: - [ ]

Comment (Squareys) [inline src/Magnum/OvrIntegration/Hmd.h]: The enum was removed in favor of return string in `Hmd::eyeCup()`

File: src/Magnum/OvrIntegration/Context.h (hunk 2 above)
Comment (Squareys) [inline src/Magnum/OvrIntegration/Context.h]: In the SDK there is a struct of 2 bools which is called `ovrDetectResult`

Comment (Squareys) [inline src/Magnum/OvrIntegration/Context.h]: @mosra Would you be okay with keeping the naming as is? Would you suggest just returning the same structure/creating an own equivalent structure instead of using enums?

Comment (mosra) [inline src/Magnum/OvrIntegration/Context.h]: Sorry about a late reply, got busy once again ... 

Enum is great, *way* better than a structure of two bools (heh), but I feel like there is an inconsistency in the naming (the enum has `*Flag`, but the enum set doesn't). Elsewhere I was always having either `SomethingFlag` and `SomethingFlags` or tried very hard to invent something else so that is always easy to guess name of the other type. 

Hmm, what about `OvrDetectResult` for the enum and `OvrDetectResults` for the enumset?

Comment (Squareys) [inline src/Magnum/OvrIntegration/Context.h]: > Hmm, what about OvrDetectResult for the enum and OvrDetectResults for the enumset? (Yes, I know it's not consistent with ovrDetectResult.

No, that's great, very good compromise.

**Comment by Squareys** [comment]
@mosra Once again, done from my perspective. Can potentially be merged with https://github.com/mosra/magnum-examples/pull/19 :)

File: src/Magnum/OvrIntegration/CMakeLists.txt [hunk 5]
```diff
@@ -46,6 +46,10 @@ set(MagnumOvrIntegration_HEADERS
 
     visibility.h)
 
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CORRADE_CXX_FLAGS}")
+
+include_directories(${MAGNUM_INCLUDE_DIRS} ${MAGNUM_APPLICATION_INCLUDE_DIRS} ${MAGNUM_AUDIO_INCLUDE_DIRS} ${CMAKE_CURRENT_BINARY_DIR})
+
```
Comment (mosra) [inline src/Magnum/OvrIntegration/CMakeLists.txt]: Why is this change here? These operations are already done in `src/CMakeLists.txt`.

File: src/Magnum/OvrIntegration/Hmd.cpp [hunk 6]
```diff
@@ -26,6 +26,8 @@
 
 #include "Magnum/OvrIntegration/Hmd.h"
 
+#include <cstring>
+
```
Comment (mosra) [inline src/Magnum/OvrIntegration/Hmd.cpp]: I guess this is some leftover, not needed anywhere.

File: src/Magnum/OvrIntegration/Hmd.h [hunk 7]
```diff
@@ -358,24 +370,110 @@ class MAGNUM_OVRINTEGRATION_EXPORT Hmd {
...
+        }
+
+        /** @brief Eye height set in the currently active Oculus profile */
+        Float eyeHeight() const {
+            return ovr_GetFloat(_session, OVR_KEY_EYE_HEIGHT, OVR_DEFAULT_EYE_HEIGHT);
+        }
+
+        /**
+         * @brief Inter populary distance set in the currently active Oculus
+         * profile
```
Comment (mosra) [inline src/Magnum/OvrIntegration/Hmd.h]: As noted previously, we have to put `@brief`s onto single line to make Doxygen parse them properly. Also, "interpupillary", I think :)

File: src/Magnum/OvrIntegration/Compositor.h [hunk 8]
```diff
@@ -149,6 +149,35 @@ class MAGNUM_OVRINTEGRATION_EXPORT Layer {
         friend class Compositor;
 };
 
+class MAGNUM_OVRINTEGRATION_EXPORT HeadLockableLayer: public Layer {
```
Comment (mosra) [inline src/Magnum/OvrIntegration/Compositor.h]: Missing class documentation.

File: src/Magnum/OvrIntegration/Hmd.h [hunk 9]
```diff
@@ -141,14 +138,11 @@ Framebuffer framebuffer{{}, textureSize};
...
 
 // setup depth attachment
-Image2D image(PixelFormat::DepthComponent, PixelType::UnsignedInt, textureSize, nullptr);
-
 Texture2D* depth = new Texture2D();
 depth->setMinificationFilter(Sampler::Filter::Linear)
       .setMagnificationFilter(Sampler::Filter::Linear)
       .setWrapping(Sampler::Wrapping::ClampToEdge)
       .setStorage(1, TextureFormat::DepthComponent24, textureSize)
-      .subImage(0, {{}, textureSize}, image);
```
Comment (mosra) [inline src/Magnum/OvrIntegration/Hmd.h]: The `;` should be on previous line then :)

File: src/Magnum/OvrIntegration/Test/EnumTest.cpp [hunk 10]
```diff
@@ -115,6 +133,26 @@ void EnumTest::errorType() {
...
 }
 
+void EnumTest::ovrDetectResult() {
+    std::ostringstream out;
+    Debug(&out) << OvrDetectResult::HmdConnected;
+    CORRADE_COMPARE(out.str(), "OvrIntegration::OvrDetectResult::HmdConnected\n");
+
+    out.str("");
+    Debug(&out) << OvrDetectResult(-1);
+    CORRADE_COMPARE(out.str(), "OvrIntegration::OvrDetectResult::(invalid)\n");
```
Comment (mosra) [inline src/Magnum/OvrIntegration/Test/EnumTest.cpp]: :+1: awesome, you are testing really everything :)

File: src/Magnum/OvrIntegration/Hmd.h [hunk 11]
```diff
@@ -358,24 +370,110 @@ class MAGNUM_OVRINTEGRATION_EXPORT Hmd {
...
         ovrPosef _ovrPoses[2];
         ovrVector3f _hmdToEyeViewOffset[2];
         ::ovrViewScaleDesc _viewScale;
 
-        ovrFrameTiming _frameTiming;
+        Double _predictedDisplayTime;
         ovrTrackingState _trackingState;
 
-        UnsignedInt _frameIndex;
+        long long _frameIndex;
```
Comment (mosra) [inline src/Magnum/OvrIntegration/Hmd.h]: Signed? Magnum has `Long`/`UnsignedLong` type which maps to `(unsigned) long long`. Shouldn't also the `incFrameIndex()` function have different return type then?

File: src/Magnum/OvrIntegration/Hmd.h [hunk 12]
```diff
@@ -358,24 +370,110 @@ class MAGNUM_OVRINTEGRATION_EXPORT Hmd {
...
+         */
+        std::array<Float, 2> maxEyeToPlateDistance() const {
+            std::array<Float, 2> values{{0.0f, 0.0f}};
+            ovr_GetFloatArray(_session, OVR_KEY_MAX_EYE_TO_PLATE_DISTANCE, values.data(), 2);
+            return values;
+        }
+
+        /** @brief Eye cup set in the currently active Oculus profile */
+        std::string eyeCup() const {
+            return ovr_GetString(_session, OVR_KEY_EYE_CUP, "A");
```
Comment (mosra) [inline src/Magnum/OvrIntegration/Hmd.h]: I guess there is some really simple explanation, but why the `"A"` here? Is that some reasonable default?

File: src/Magnum/OvrIntegration/HmdEnum.h [hunk 13]
```diff
@@ -184,6 +211,10 @@ enum class ErrorType: Int {
     MismatchedAdapters = ovrError_MismatchedAdapters,   /**< Chosen rendering adapters between client and service do not match */
     LeakingResources = ovrError_LeakingResources,       /**< Calling application has leaked resources */
     ClientVersion = ovrError_ClientVersion,             /**< Client version too old to connect to service */
+    OutOfDateOS = ovrError_OutOfDateOS,                 /**< The operating system is out of date. */
+    OutOfDateGfxDriver = ovrError_OutOfDateGfxDriver,   /**< The graphics driver is out of date. */
+    IncompatibleGPU = ovrError_IncompatibleGPU,         /**< The graphics hardware is not supported */
+    NoValidVRDisplaySystem = ovrError_NoValidVRDisplaySystem,   /**< No valid VR display system found. */
```
Comment (mosra) [inline src/Magnum/OvrIntegration/HmdEnum.h]: `IncompatibleOs`, `OutOfDateOs`, `IncompatibleGpu`, `NoValidVrDisplaySystem`, `UsbBandwidth`, `UsbEnumeratedSpeed` and `HmdFirmwareMismatch` so we have everything CamelCase to be consistent with the rest of Magnum. Looks like even Oculus can't get it consistent in a single enum :)

File: src/Magnum/OvrIntegration/CMakeLists.txt (hunk 5 above)
Comment (Squareys) [inline src/Magnum/OvrIntegration/CMakeLists.txt]: Ah, I missed those. I was really surprised how this missing made it to code review, should have investigated further, I guess. Will remove it again, sorry ;)

File: src/Magnum/OvrIntegration/Hmd.h (hunk 9 above)
Comment (Squareys) [inline src/Magnum/OvrIntegration/Hmd.h]: oops :P Good catch.

File: src/Magnum/OvrIntegration/Test/EnumTest.cpp (hunk 10 above)
Comment (Squareys) [inline src/Magnum/OvrIntegration/Test/EnumTest.cpp]: Yeah, that invalid thing is my most frequent source of copy-paste error, so back then I started doing that for all of the many OvrIntegration enums.

File: src/Magnum/OvrIntegration/Hmd.h (hunk 11 above)
Comment (Squareys) [inline src/Magnum/OvrIntegration/Hmd.h]: Ah, I see, thanks for pointing that out. Yes, `incFrameIndex` should probably return the same type.

File: src/Magnum/OvrIntegration/Hmd.h (hunk 12 above)
Comment (Squareys) [inline src/Magnum/OvrIntegration/Hmd.h]: I guess that is the default, and I probably took that from some defaults by the SDK itself, I'll check again, though, just to be on the safe side (in my oppinion "B" would be a more reasonable default, since it's the lense without sight correction)

**Comment by Squareys** [comment]
@mosra Hey, you're back! :D I will take some time on the changes you proposed, probably wont find time before the weekend.

**Comment by Squareys** [comment]
@mosra Alright, I finally got around to taking care of the flaws you pointed out! :)

**Comment by mosra** [comment]
:+1: Merged.

I just rebased the commits to current master and did a single change -- removed redundant prefix from the `OvrDetectResult` enum. Hope that's okay.

**Comment by Squareys** [comment]
> Hope that's okay.

No problem, of course, thanks!
